src/rust-crypto/verification.ts

```diff
@@ -477,19 +483,39 @@ abstract class BaseRustVerifer<InnerType extends RustSdkCryptoJs.Qr | RustSdkCry
 
                 this.emit(VerificationRequestEvent.Change);
             };
-            inner.registerChangesCallback(onChange);
+            inner.registerChangesCallback(this.onChangeCallback);
         });
         // stop the runtime complaining if nobody catches a failure
         this.completionPromise.catch(() => null);
     }
 
+    /**
+     * Replace the inner Rust verifier with a different one.
+     *
+     * @param inner - the new Rust verifier
+     * @internal
+     */
+    public setInner(inner: InnerType): void {
+        if (this.inner != inner) {
+            this.inner = inner;
+            inner.registerChangesCallback(this.onChangeCallback);
+            this.onSetInner();
+            this.onChangeCallback();
+        }
+    }
+
     /**
      * Hook which is called when the underlying rust class notifies us that there has been a change.
      *
      * Can be overridden by subclasses to see if we can notify the application about an update.
      */
     protected onChange(): void {}
 
+    /**
+     * Hook which is called when the underlying rust class is changed by calling `setInner`.
+     */
+    protected onSetInner(): void {}
+
```

Comment:
I rather imagined that you would move ~~`setInner`~~ `replaceInner` into `RustSASVerifier`. All these callback hooks feel complicated.